src/server/graceful.rs

```diff
@@ -0,0 +1,620 @@
+//! Utility to gracefully shutdown a server.
+//!
+//! This module provides a [`GracefulShutdown`] type,
+//! which can be used to gracefully shutdown a server.
+//!
+//! See <https://github.com/hyperium/hyper-util/blob/master/examples/server_graceful.rs>
+//! for an example of how to use this.
+
+use pin_project_lite::pin_project;
+use slab::Slab;
+use std::{
+    fmt::{self, Debug},
+    future::Future,
+    pin::Pin,
+    sync::{atomic::AtomicUsize, Arc, Mutex},
+    task::{self, Poll, Waker},
+};
+
+/// A graceful shutdown watcher
+pub struct GracefulShutdown {
+    // state used to keep track of all futures that are being watched
+    state: Arc<GracefulState>,
+    // state that the watched futures to know when shutdown signal is received
+    future_state: Arc<GracefulState>,
+}
+
+impl Debug for GracefulShutdown {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("GracefulShutdown").finish()
+    }
+}
+
+impl Default for GracefulShutdown {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GracefulShutdown {
+    /// Create a new graceful shutdown watcher
+    pub fn new() -> Self {
+        Self {
+            state: Arc::new(GracefulState {
+                counter: AtomicUsize::new(0),
+                waker_list: Mutex::new(Slab::new()),
+            }),
+            future_state: Arc::new(GracefulState {
+                counter: AtomicUsize::new(1),
```

Comment:
It's because this is the property used by the tasks to know if a graceful cancellation is requested. As such it starts at 1 as the main struct is still not cancelled. It goes to 0 once it is being cancelled. It looks a bit odd but it's basically reusing the same `GracefulState` struct that was already used for the other direction